src/desktop/space.rs

```diff
@@ -234,8 +241,124 @@ impl Space {
         self.outputs.get(0).cloned()
     }
 
-    pub fn cleanup(&mut self) {
+    pub fn refresh(&mut self) {
         self.windows.retain(|w| w.toplevel().alive());
+
+        for output in &mut self.outputs {
+            output_state(self.id, output)
+                .surfaces
+                .retain(|s| s.as_ref().is_alive());
+        }
+
+        for window in &self.windows {
+            let bbox = window_rect(window, &self.id);
+            let kind = window.toplevel();
+
+            for output in &self.outputs {
+                let output_geometry = self
+                    .output_geometry(output)
+                    .unwrap_or_else(|| Rectangle::from_loc_and_size((0, 0), (0, 0)));
+                let mut output_state = output_state(self.id, output);
+
+                // Check if the bounding box of the toplevel intersects with
+                // the output, if not no surface in the tree can intersect with
+                // the output.
+                if !output_geometry.overlaps(bbox) {
+                    if let Some(surface) = kind.get_surface() {
+                        with_surface_tree_downward(
+                            surface,
+                            (),
+                            |_, _, _| TraversalAction::DoChildren(()),
+                            |wl_surface, _, _| {
+                                if output_state.surfaces.contains(wl_surface) {
+                                    slog::trace!(
+                                        self.logger,
+                                        "surface ({:?}) leaving output {:?}",
+                                        wl_surface,
+                                        output.name()
+                                    );
+                                    output.leave(wl_surface);
+                                    output_state.surfaces.retain(|s| s != wl_surface);
+                                }
+                            },
+                            |_, _, _| true,
+                        )
+                    }
+                    continue;
+                }
+
+                if let Some(surface) = kind.get_surface() {
+                    with_surface_tree_downward(
```

Comment:
I'm wondering if it makes sense to give enter/leave information at the subsurface granularity detail.

Wouldn't it be simpler to just always send the same enter/leave to the whole subsurface tree, based only on intersection with the bounding box? I seems to me that it's unnecessary detail computed and sent to the client.

Reply:
Hmm. Not sure actually. While that would probably be fine for most clients, I am not sure if more complex clients like firefox would agree?? But potentially worth a try? This is at least what anvil has been doing.

Reply:
Does the wayland core protocol say something about this?

Reply:
Not much, tbh. The information of enter/leave is mostly used for deciding the scaling factor when drawing, the client cannot assume that its surface is actually visible on all the wl_outputs it is "entered" (as it might be hidden behind an other for example).

So I don't think it's actually possible for a client to rely on precise tracking of the wl_output its surface are entered on.